torch/csrc/api/include/torch/nn/modules/sequential.h

```diff
@@ -347,16 +356,54 @@ class SequentialImpl : public Cloneable<SequentialImpl> {
   /// The base case, when the list of modules is empty.
   void push_back() {}
 
-  // Box the AnyModules to give Sequential reference semantics, like the rest of
-  // the API. Note that this is not required otherwise, this could just be a
+  // Box the AnyModules to give AnySequential reference semantics, like the rest
+  // of the API. Note that this is not required otherwise, this could just be a
   // `vector<AnyModule>`.
   std::vector<AnyModule> modules_;
 };
 
+/// A `ModuleHolder` subclass for `AnySequentialImpl`.
+/// See the documentation for `AnySequentialImpl` class to learn what methods it
+/// provides, or the documentation for `ModuleHolder` to learn about PyTorch's
+/// module storage semantics.
+TORCH_MODULE(AnySequential);
+
+/// `Sequential` is derived from `AnySequential` and has a non-template
+/// returning type allowing `Sequential`s to be nested in each other making them
+/// compatible with the Python API. The template input type helps warn users
+/// about the change in the API.
+class SequentialImpl : public AnySequentialImpl {
+ public:
+  using AnySequentialImpl::AnySequentialImpl;
+
+  template <typename ReturnType = Tensor, typename... InputTypes>
+  Tensor forward(InputTypes&&... inputs) {
+    static_assert(
+        std::is_same<ReturnType, Tensor>(),
+        "Can only call Sequential::forward with Tensor as output type."
+        " If you would like to have a non-Tensor output type, please use AnySequential instead.");
+    return AnySequentialImpl::forward(std::forward<InputTypes>(inputs)...);
+  }
+
+  void pretty_print(std::ostream& stream) const override {
+    stream << "torch::nn::Sequential";
+  }
+};
+
 /// A `ModuleHolder` subclass for `SequentialImpl`.
 /// See the documentation for `SequentialImpl` class to learn what methods it
 /// provides, or the documentation for `ModuleHolder` to learn about PyTorch's
 /// module storage semantics.
-TORCH_MODULE(Sequential);
+
+class Sequential : public torch::nn::ModuleHolder<SequentialImpl> {
+ public:
+  using torch::nn::ModuleHolder<SequentialImpl>::ModuleHolder;
+
+  Sequential(std::shared_ptr<AnySequentialImpl> any_impl) {
+    for (auto& module : *any_impl)
+      (*this)->push_back(std::move(module));
```

Comment:
@yf225 It fails to compile here because ```void push_back(AnyModule any_module)``` is private within ```AnySequentialImpl```. Am I doing something wrong or this function should be public?

Reply:
Would `std::shared_ptr<Module> operator[](size_t index)` or `std::shared_ptr<Module> ptr(size_t index)` within `AnySequentialImpl` work for getting individual submodules? With them we should be able to use the public `push_back` APIs such as `void push_back(std::string name, std::shared_ptr<ModuleType> module_ptr)`.

As a side note, we would also need to make sure the sequential submodule's names are preserved (by adding a test that checks it).

Reply:
@yf225 actually that won't work. because ```Module``` doesn't have ```forward``` function and the code fails to compile:
```
no member named 'forward' in 'torch::nn::Module'
          &std::remove_reference<ModuleType>::type::forward)) {
```

The code:
```
  Sequential(std::shared_ptr<AnySequentialImpl> module) {
    auto named = module->named_modules();
    for (size_t i = 0; i < named.size(); ++i)
      (*this)->push_back(named[i].key(), named[i].value());
  }
```

Reply:
@ShahriarSS I think writing a constructor for `SequentialImpl` that takes `AnySequentialImpl` is a great idea. Thanks a lot for working on it!